wsmux/session.go

```diff
@@ -0,0 +1,237 @@
+package wsmux
+
+import (
+	"encoding/binary"
+	// "io"
+	"io/ioutil"
+	// "log"
+	"net"
+	// "runtime"
+	"bytes"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	// "github.com/taskcluster/slugid-go/slugid"
+)
+
+const (
+	defaultQueueSize       = 10
+	defaultStreamQueueSize = 10
+)
+
+type Session struct {
+	mu       sync.Mutex
+	streams  map[uint32]*stream
+	streamCh chan *stream
+	writes   chan frame
+	conn     *websocket.Conn
+
+	acceptDeadline time.Time
+	readDeadline   time.Time
+	writeDeadline  time.Time
+
+	logger Logger
+
+	closed       atomic.Value
+	remoteClosed atomic.Value
+
+	nextID uint32
+
+	remoteCloseCallback func()
+}
+
+func newSession(conn *websocket.Conn, server bool, conf *Config) *Session {
+	if conf == nil {
+		conf = &Config{}
+	}
+	s := &Session{}
+	s.conn = conn
+	s.streams = make(map[uint32]*stream)
+	s.streamCh = make(chan *stream, defaultStreamQueueSize)
+	s.writes = make(chan frame, defaultQueueSize)
+
+	s.acceptDeadline = conf.AcceptDeadline
+	s.readDeadline = conf.ReadDeadline
+	s.writeDeadline = conf.WriteDeadline
+	s.remoteCloseCallback = conf.RemoteCloseCallback
+
+	s.closed.Store(false)
+	s.remoteClosed.Store(false)
+
+	if server {
+		s.nextID = 0
+	} else {
+		s.nextID = 1
+	}
+
+	if conf.Log == nil {
+		s.logger = &nilLogger{}
+	} else {
+		s.logger = conf.Log
+	}
+	go s.sendLoop()
+	go s.recvLoop()
+	return s
+}
+
+func (s *Session) Accept() (net.Conn, error) {
+	var timeout <-chan time.Time
+	var timer *time.Timer
+	if !s.acceptDeadline.IsZero() {
+		timer = time.NewTimer(s.acceptDeadline.Sub(time.Now()))
+		timeout = timer.C
+	}
+
+	select {
+	case <-timeout:
+		return nil, ErrAcceptTimeout
```

Comment:
This is another place where reading from `s.done` makes sense...

Also the timeouts, wouldn't it be better if those were timers on either `Session` object... Okay, probably you have to find out how they are used :)